### ISSUE TITLE
Building an F# 3.0 (FSharp.Core, 2.3.0.0) app fails

### ISSUE THREAD

**Issue by atifaziz** [Microsoft/visualfsharp#4627]
#### Repro steps

Provide the steps required to reproduce the problem

1. In Visual Studio 2017, select **File** &rarr; **New** &rarr; **Project**
2. When **New Project** dialog appears, select **Installed** &rarr; **Visual F#** in the tree pane on the left and then choose **Console Applicaiton**
3. From the **Framework** drop-down, select **.NET Framework 2.0** as shown below:
   ![image](https://user-images.githubusercontent.com/20511/37955496-06b81b16-31a9-11e8-87d1-99c04b2599cf.png)
4. Click **OK**.

#### Expected behavior

Expected the new project to be created and build without errors.

If F# 3.0 (FSharp.Core, 2.3.0.0) or building for .NET Framework 2.0 is no longer supported then I expect the UI options to be disabled or absent. If additional (out-of-band) components need to be downloaded and installed then the UI should reflect that.

#### Actual behavior

After step 4, the following error appears:

![image](https://user-images.githubusercontent.com/20511/37955511-17185804-31a9-11e8-926c-96bc70009fae.png)

Once the error dialog box is closed and one tries to build the project, the build fails with the following output:

```
1>------ Build started: Project: TestApp, Configuration: Debug Any CPU ------
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2052,5): warning MSB3267: The primary reference "System.ValueTuple", which is a framework assembly, could not be resolved in the currently targeted framework. ".NETFramework,Version=v2.0". To resolve this problem, either remove the reference "System.ValueTuple" or retarget your application to a framework version which contains "System.ValueTuple".
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2052,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "FSharp.Core". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors.
1>FSC: error FS0193: The module/namespace 'System' from compilation unit 'mscorlib' did not contain the namespace, module or type 'Lazy`1'
1>T:\TestApp\Program.fs(4,3): error FS0039: The type 'EntryPoint' is not defined.
1>T:\TestApp\Program.fs(6,5): error FS0039: The value or constructor 'printfn' is not defined.
1>Done building project "TestApp.fsproj" -- FAILED.
1>
1>Build FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

#### Additional information 

* Operating system: Microsoft Windows 10 Pro, version 10.0.16299
* Microsoft Visual Studio Enterprise 2017 , Version 15.6.4
* Microsoft .NET Framework Version 4.7.02556 (4.7.1)

**F# desktop language support** and **.NET Framework 3.5 development tools** are installed:

![image](https://user-images.githubusercontent.com/20511/37957801-27fd38aa-31af-11e8-8276-644fcc95a7c0.png)

**Comment by dsyme** [comment]
@atifaziz You can probably workaround by one or both of 
1. adding an explicit reference to FSHarp.Core in the project file
2. disabling the implicit reference to System.ValueType in the project file

**Comment by atifaziz** [comment]
@dsyme Thanks. Removing the reference to **System.ValueType** helped somewhat. As for **FSharp.Core**, I see this in the project file:

```xml
    <Reference Include="FSharp.Core">
      <Name>FSharp.Core</Name>
      <AssemblyName>FSharp.Core.dll</AssemblyName>
      <HintPath>$(MSBuildProgramFiles32)\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\$(TargetFSharpCoreVersion)\FSharp.Core.dll</HintPath>
    </Reference>
```

where `TargetFSharpCoreVersion` is defined as 2.3.0.0 earlier:

```xml
    <TargetFSharpCoreVersion>2.3.0.0</TargetFSharpCoreVersion>
```

Peeking into the references directory, I don't see anything for that version:

```
    Directory: C:\Program Files (x86)\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0


Mode                LastWriteTime         Length Name                                                                                                     
----                -------------         ------ ----                                                                                                     
d-----        3/14/2018   3:18 PM                4.3.0.0                                                                                                  
d-----        3/14/2018   3:18 PM                4.3.1.0                                                                                                  
d-----        3/14/2018   3:18 PM                4.4.0.0                                                                                                  
d-----        3/14/2018   3:18 PM                4.4.1.0                                                                                                  
d-----        3/14/2018   3:18 PM                4.4.3.0                                                                                                  
```

Any clues as to where I'm supposed to find version 2.3.0.0 on disk so I can add the explicit reference?

**Comment by dsyme** [comment]
@atifaziz Try the old versions of the nuget package https://www.nuget.org/packages/FSharp.Core/3.1.2.5

**Comment by cartermp** [comment]
@atifaziz Any success with the older NuGet version?

**Comment by atifaziz** [comment]
@cartermp @dsyme Not really. Oddly, I got the following error:

    Could not install package 'FSharp.Core 3.1.2'. You are trying to install this package into a project that targets '.NETFramework,Version=v2.0', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.

Looking into the package, though, I can clearly see `FSharp.Core.dll` for `net20`:

![image](https://user-images.githubusercontent.com/20511/44090354-32f1a08c-9fca-11e8-8cd9-dbcb628d0633.png)

---
Since reporting this issue, my environment has changed as follows:

- Operating system: Microsoft Windows 10 Pro, version 10.0.17134.112
- Microsoft Visual Studio Enterprise 2017 , Version 15.7.2
- Microsoft .NET Framework Version 4.7.03056 (4.7.2)

**Comment by cartermp** [comment]
Hmmm, that is odd. If you manually include that binary by editing the project file, does it still work? Perhaps there is a quirk in some intersection of F#/FSharp.Core/net20/NuGet.

**Comment by atifaziz** [comment]
Ok, my bad. I noticed that @dsyme specifically called out [version 3.1.2.5](https://www.nuget.org/packages/FSharp.Core/3.1.2.5) whereas in [the error message I quoted earlier](https://github.com/Microsoft/visualfsharp/issues/4627#issuecomment-412847935), you can see I installed [version 3.1.2](https://www.nuget.org/packages/FSharp.Core/3.1.2):

    Could not install package 'FSharp.Core 3.1.2'. You are trying to install this package into a project that targets '.NETFramework,Version=v2.0', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.

In the [NuGet Package Explorer screenshot](https://user-images.githubusercontent.com/20511/44090354-32f1a08c-9fca-11e8-8cd9-dbcb628d0633.png), I was looking at 3.1.2.5. Apparently, the revision number of the package made all the different. Whereas 3.1.2 didn't have any assemblies for anything below .NET 4.0, 3.1.2.5 did. Perhaps that was the patch to that package.

So installing [version 3.1.2.5](https://www.nuget.org/packages/FSharp.Core/3.1.2.5) does work and allow the program to compile. :+1:

**Comment by cartermp** [comment]
Aha! That'd do it I suppose 🙂. I'll close this out for now, but please do ping back if you notice any further issues.

Another potential work item here is to update the [docs](https://docs.microsoft.com/en-us/dotnet/fsharp/get-started/install-fsharp?tabs=windows) to mention how to get FSharp.Core for older .NET Framework targets. Does that seem like a good place to have that information?

**Comment by atifaziz** [comment]
> Does that seem like a good place to have that information?

It can go there but frankly having **File** &rarr; **New Project** fail doesn't seem right. Either that should be fixed (by using the right version of FSharp.Core package) or choosing an older framework that requires manual setup should lead to [the documentation](https://docs.microsoft.com/en-us/dotnet/fsharp/get-started/install-fsharp?tabs=windows) where one can follow the steps. That said, thanks @dsyme & @cartermp for working with me to get to the bottom of this, even if it took months. Fortunately, it wasn't time-critical for me. :wink:

**Comment by cartermp** [comment]
@KevinRansom would have to speak to how difficult something like that could be, but we only recently switched to consuming the package with 15.8. Your repros are both with older VS versions where these templates select a binary deployed on disk rather than the NuGet package.

Another issue is that we build against net45 for .NET Framework projects, which complicates matters. This motivated #5509 where we do not show older .NET Framework versions in the dropdown, meaning that to target these older versions, people will have to manually set the right things in their new project files. Older, existing projects will still work, but new ones will require this workaround.

**Comment by KevinRansom** [comment]
@artifaziz not limiting the framework to greater than or equal to 4.5 when creating a new project was a bug.   That wierdly enough I fixed yesterday.

The nuget package FSharp.Core 4.1.18 ships an FSharp.Core 2.3.0.0.  However, the latest Visual FSharp Tools are not expected to support building .NET Framework apps targetting a framework lower than 4.5.  It may work, it may not, we have certainly done nothing to ensure it works.

I'm sorry that the new isn't better.

Kevin
